builder/dockerfile/builder.go

```diff
@@ -322,6 +325,10 @@ func (b *Builder) dispatchDockerfileWithCancellation(parseResult []instructions.
 			if err := dispatch(dispatchRequest, cmd); err != nil {
 				return nil, err
 			}
+			if b.options.Remove && !b.options.ForceRemove {
```

Comment:
There is another `ForceRemove` defer inside the `dispatch` function https://github.com/moby/moby/blob/d9cd40d58bef2f607d4a798c99eb98cf56aab112/builder/dockerfile/evaluator.go#L56 that overrides this condition. I think we should keep always removing after each dispatch to be consistent with old versions and better protect against crashes. This is how it works atm as well but this code/comment indicates the opposite.

Reply:
Good catch @tonistiigi. I'll put remove handling in the global dispatch function.